assets/app/scripts/controllers/builds.js

```diff
@@ -76,8 +76,13 @@ angular.module('openshiftConsole')
         if (build.metadata.labels) {
           buildConfigName = build.metadata.labels.buildconfig || "";
         }
-        $scope.buildsByBuildConfig[buildConfigName] = $scope.buildsByBuildConfig[buildConfigName] || {};
-        $scope.buildsByBuildConfig[buildConfigName][buildName] = build;
+        // Either we are filtering and the build config exists and isn't filtered or the build with no build config is matched by the selector
+        // or we aren't filtering
+        var labelSelector = LabelFilter.getLabelSelector();
+        if ((!labelSelector.isEmpty() && ($scope.buildConfigs[buildConfigName] || labelSelector.matches(build))) || LabelFilter.getLabelSelector().isEmpty()) {
```

Comment:
I feel like this `if` statement is complicated enough to deserve its own function like

```
function showBuild(build) {
  var labelSelector = LabelFilter.getLabelSelector(), buildConfigName;
  if (labelSelector.isEmpty() {
    return true;
  }

  buildConfigName = build.metadata.labels.buildconfig;
  if (buildConfigName) {
   return $scope.buildConfigs[buildConfigName];
  }

  return labelSelector.matches(build));
}

...

if (showBuild(build)) { ... }
```